### ISSUE TITLE
When I train with yolov8, the image size is 640*640, and the actual scene detected is an 8k camera, is it so scientific, what should I do?

### ISSUE THREAD

**Issue by threeneedone** [ultralytics/ultralytics#2078]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

When I train with yolov8, the image size is 640*640, and the actual scene detected is an 8k camera, is it so scientific, what should I do?

**Comment by yoloyash** [comment]
you should look into slicing aided inference with [sahi](https://github.com/obss/sahi/blob/main/demo/inference_for_yolov8.ipynb)

**Comment by Laughing-q** [comment]
@threeneedone depends on what's the ratio of `the size of objects / the whole image`. If the ratio is not that small which means most of your objects are large then I think it's fine to try a training with 640x640, but if ratio is small and the objects in your scene are small then I recommend you to use larger image-size, i.e 1280, but this would cost more training resources.
And yes `sahi` could be a solution for you if your objects are so small.

**Comment by xskopa03** [comment]
Is there any alternative to sahi which support C++?

**Comment by glenn-jocher** [comment]
@xskopa03 yes, there are several alternatives to Sahi that support C++. One of them is OpenCV, which is a popular computer vision library that has many built-in functions that can perform object detection. It provides a pre-trained object detection model and different options are available to fine-tune the same as per the need.

Another alternative is the Intel OpenVINO toolkit, which is a high-performance computer vision library that provides support for different programming languages such as C++, Python, and others. It has a pre-trained model zoo that includes object detection models such as YOLOv3 and YOLOv4, and it also provides tools to optimize and deploy these models on different devices. 

Besides OpenCV and OpenVINO, there are other libraries and frameworks such as Tensorflow, Darknet, and Caffe, which also support C++ and provide pre-trained object detection models.

**Comment by xskopa03** [comment]
@glenn-jocher I already use  C++ TensorRT engine to inference yolo, but I have problems with detection of small objects.  I was thinking about the slicing the input image into tiles and then run inference on them and do a NMS. Thats what the sahi project has already implemented in Python. So I was asking if there is any C++ alternative which has this feature to use.  I did not find any :/ So I guess  I will have to implement by myself, use overlapping tiles and then use NMS to match overlapping tiles, but this would detect only object that are small or big as a tile. I do not know what is a right algorithm for this to achieve. Maybe run  inference on tiles and then on whole picture?

**Comment by glenn-jocher** [comment]
@xskopa03, using Sahi would affect the speed of YOLOv8 in the detection process, as the process of slicing the image into smaller tiles and then merging them using non-maximum suppression (NMS) would require additional processing overhead. However, this method can help increase the accuracy of detecting small objects in the scene. The size of the sliding window should be specified based on the size of objects in the scene, and it's not necessarily required to be the same as the size of the images used in the training process. It's important to perform experiments and find the optimal configuration for your specific use case. 

Regarding the release of the yolov8_p6 detection network, I don't think there's been an official announcement about its release yet. However, you can stay up-to-date with Ultralytics YOLOv8 by checking their Github repository or their official website.

**Comment by derekboase** [comment]
> @derekboase yes, using Sahi to slice your image into smaller tiles and then merging them after detection will add more computation and thus could potentially affect the inference speed of your model. The size of the sliding window in Sahi is configurable and it's not necessary to be the same as the size of the images used for training your model.
> 
> You would typically adjust the size of the sliding window based on the size of the objects you expect to detect in your images. If your objects of interest are smaller, you might want a smaller window size, and larger window size for larger objects. However, it's always best to experiment with different sizes to find what works best for your specific use case.

@glenn-jocher, thank you for the timely reply. I notice that you are incredibly active, helping people out in almost every discussion I've come across, I appreciate your hard work and useful insights.

I'm looking to compare performance of the yolov8 models to the models I've been using in TFOD api from [https://github.com/tensorflow/models/blob/master/research/object_detection/g3doc/tf2_detection_zoo.md](model zoo) and I have high resolution images that contain small, medium and large objects. My thoughts were to train two or three models for each size of object (found using k-means on the average size and aspect ratio) then slice the images for the smaller objects and train the yolov8 model on the sliced images so that I could use higher resolution images (maybe 2000, 1500 size if computation allows). Is there a way I can use sahi in the training process, or is it strictly for inferencing?

**Comment by glenn-jocher** [comment]
@derekboase hi Derek,

I'm glad to hear that you found our responses to be helpful! 

Your approach of training separate models for different object sizes sounds promising. Keep in mind that it is not typical to use Sahi during the training phase, primarily because Sahi is designed as an inferencing tool - it helps by breaking the scene down into manageable pieces so that the neural network can make predictions more effectively in scenes with high-resolution or complex backgrounds.

For the training phase, the common approach is to adjust the input image size to match the sizes of the objects. To take advantage of the full resolution of your images, you might prefer to use higher resolution as the input size of YOLOv8 during the training phase, as you've suggested. By doing so, the YOLOv8 model will be able to effectively learn features from small, medium, and large objects in different scales of images. With this method, after training, you'll have a model that can detect objects of different sizes in a high-resolution scene.

To sum up, Sahi can be applied during the inference phase to handle high-resolution images or to improve the performance on small objects, but during the training phase, the recommended way is to adjust the input size according to the sizes of objects in your dataset.

I hope this response clarifies your query. Let me know if you have any more questions!

**Comment by jpers1** [comment]
> @xskopa03 yes, there are several alternatives to Sahi that support C++. One of them is OpenCV, which is a popular computer vision library that has many built-in functions that can perform object detection. It provides a pre-trained object detection model and different options are available to fine-tune the same as per the need.
> 
> Another alternative is the Intel OpenVINO toolkit, which is a high-performance computer vision library that provides support for different programming languages such as C++, Python, and others. It has a pre-trained model zoo that includes object detection models such as YOLOv3 and YOLOv4, and it also provides tools to optimize and deploy these models on different devices.
> 
> Besides OpenCV and OpenVINO, there are other libraries and frameworks such as Tensorflow, Darknet, and Caffe, which also support C++ and provide pre-trained object detection models.

SAHI is a tool to slice high resolution images and wrap low-resolution object detection, then merge the results on a sliced into single representation which is equivalent, but better quality, than if whole image would just been resized and feed into the trained object detector.

I doubt that all tools/frameworks that you posted here provide this functionality, so please be more specific (I am interested in problems SAHI adresses as well) because your answer is incredibly generic, to the point that I suspect it may be basically wrong and misleading.

**Comment by glenn-jocher** [comment]
@jpers1 apologies for any confusion earlier. To address your specific concern, if you're looking for a C++ tool that replicates Sahi's functionality for slicing high-resolution images for object detection and then merging the results, I'm not aware of a direct alternative that offers this out-of-the-box.

For YOLOv8, you might need to implement a custom solution in C++ that involves dividing the high-resolution image into overlapping tiles, running inference on each tile, and then applying a custom NMS across the tile results to merge detections. This approach would allow you to detect objects that are smaller or as large as the tile itself.

As for the release of a YOLOv8_p6 network, we don't have information on that at the moment. Keep an eye on the Ultralytics GitHub repository for any updates on new model releases.

**Comment by CyVision13** [comment]
@glenn-jocher 
![image](https://github.com/ultralytics/ultralytics/assets/44062995/851e4dfd-1f42-43b8-b5a7-7029846419c2)
Hi i used SAHI and yolov8 in my Project, but it selects 1 object twice because of slicing the image is there any solution for fixing this problem?

**Comment by glenn-jocher** [comment]
@CyVision13 when training a YOLOv8 model with an image size of 640x640, but deploying it in a real-world scenario with an 8K camera, there are a few considerations to keep in mind:

1. **Resolution Mismatch**: The significant difference in resolution between the training data and the deployment scenario can affect the model's performance. Objects that appear small in an 8K image may not be well-represented in the 640x640 training images.

2. **Aspect Ratio**: If the aspect ratio of the 8K camera images differs from the 640x640 training images, this could also impact the model's ability to generalize.

3. **Computational Resources**: Processing 8K images requires significantly more computational power and memory, which may not be feasible in real-time applications.

To address these issues, you can consider the following approaches:

- **Downsampling**: You could downsample the 8K images to a lower resolution that matches the training size before running inference. This would make the object sizes and image characteristics more similar to the training data.

- **Sliding Window**: If maintaining high resolution is crucial, you could apply a sliding window approach where you run the model on cropped sections of the 8K image. This would require additional logic to handle overlapping detections and reconstruct the full image's detections.

- **Retraining**: Ideally, you should retrain the model using higher-resolution images that are more representative of the 8K deployment scenario. You can resize your high-resolution images to a larger input size that the model can handle, such as 1024x1024 or higher, depending on your computational resources.

- **Multi-Scale Training**: YOLOv8 supports multi-scale training, which can help the model generalize better to different image sizes. You can enable this feature during training to expose the model to a variety of scales.

- **Fine-Tuning**: If retraining from scratch is not an option, you could fine-tune the pre-trained model on a dataset of images that are cropped or resized from the 8K camera. This would help the model adapt to the characteristics of the higher-resolution images.

In summary, while training on 640x640 images and deploying on 8K images is not ideal, there are strategies you can employ to improve the model's performance in high-resolution scenarios. Consider the trade-offs between resolution, computational resources, and real-time requirements to choose the best approach for your application.

**Comment by CyVision13** [comment]
@glenn-jocher Thanks, i will try the approaches.
i did not know that yolov8 supports multi-scale training!

**Comment by glenn-jocher** [comment]
@CyVision13 when training a YOLOv8 model with an image size of 640x640, but deploying it in a real-world scenario with an 8K camera, there are a few considerations to keep in mind:

1. **Resolution Mismatch**: The significant difference in resolution between the training images and the deployment images can affect the model's ability to generalize. Objects in the 8K images may appear much smaller or larger than those the model was trained on.

2. **Aspect Ratio**: If the aspect ratio of the 8K images differs from the 640x640 training images, this could also impact the model's performance.

3. **Computational Resources**: Processing 8K images requires significantly more computational power and memory, which may not be feasible in real-time applications.

To address these issues, you can consider the following approaches:

- **Downsampling**: You can downsample the 8K images to a lower resolution closer to the training size before running them through the model. This will make the object sizes more consistent with the training data.

- **Sliding Window or Cropping**: If downsampling results in objects being too small to detect, you can use a sliding window or cropping approach to process the 8K image in sections. This allows you to maintain a higher resolution while still fitting the image into the model's expected input size.

- **Fine-tuning**: If possible, fine-tune the model on a dataset that includes higher-resolution images or images that are more representative of the deployment scenario. This can help the model adapt to the characteristics of the 8K images.

- **Multi-Scale Training**: Train the model using a range of image sizes that includes both the original training size and larger sizes that are closer to the deployment resolution. This can help the model learn to detect objects across different scales.

- **Model Input Size**: If computational resources allow, you can train the model with a larger input size that's more representative of the 8K resolution. However, this will require more memory and computational power.

- **Post-Processing**: Adjust post-processing steps like non-maximum suppression (NMS) thresholds to account for the change in object sizes due to the resolution difference.

In summary, while training on 640x640 images and deploying on 8K images is not ideal, there are strategies you can employ to bridge the gap and improve the model's performance in the real-world scenario. It's important to experiment with these approaches to find the best solution for your specific use case.

**Comment by tiffany-sats** [comment]
hi @glenn-jocher! 

1. when i train with yolov8, the training image sizes are reduced to 640*640. however, during actual deployment, if my CCTV is 4K resolution, does the yolo model reduce the image resolution to 640*640 during inferencing as well?
2. are there other yolo models that can train images on a higher resolution? as we are planning to do OCR, therefore resolution is important

**Comment by glenn-jocher** [comment]
Hi @tiffany-sats,

Great questions! Let's address them one by one:

1. **Inference Resolution**: Yes, during inference, YOLO models typically resize the input images to the size they were trained on (e.g., 640x640). This ensures consistency in the model's performance. However, you can adjust the inference resolution by specifying the `imgsz` parameter in the `predict` method, which allows you to use higher resolutions if needed. Keep in mind that higher resolutions will require more computational resources.

2. **Training with Higher Resolutions**: YOLOv8 supports training with different image sizes. You can specify a higher resolution during training by setting the `imgsz` parameter in the `train` method. For example, to train with 1024x1024 images, you can use:
   ```python
   model.train(data='data.yaml', epochs=100, imgsz=1024)
   ```
   The maximum resolution you can use depends on your hardware capabilities, particularly the GPU memory. For OCR tasks, higher resolutions can indeed be beneficial to capture finer details.

Feel free to experiment with different resolutions to find the optimal balance between performance and computational efficiency. If you encounter any issues, please provide a minimum reproducible example to help us assist you better: [Minimum Reproducible Example](https://docs.ultralytics.com/help/minimum_reproducible_example).

Best of luck with your project! 🚀

**Comment by tiffany-sats** [comment]
hi @glenn-jocher can u give me an example on how to adjust the inference resolution? is in it the YOLO("") line?

**Comment by tiffany-sats** [comment]
File "/anaconda/envs/azureml_py38/lib/python3.8/site-packages/ultralytics/utils/callbacks/raytune.py", line 17, in on_fit_epoch_end
    if ray.tune.is_session_enabled():
AttributeError: module 'ray.tune' has no attribute 'is_session_enabled'

i get this error when i train my model. any idea how to fix this?

**Comment by glenn-jocher** [comment]
Hi @tiffany-sats,

Thank you for reaching out! The error you're encountering indicates that the `ray.tune` module does not have the `is_session_enabled` attribute. This could be due to a version mismatch or an outdated package.

Here are a few steps to help resolve this issue:

1. **Update Packages**: Ensure you have the latest versions of all relevant packages. You can update `ray` and `ultralytics` using the following commands:
   ```bash
   pip install --upgrade ray ultralytics
   ```

2. **Reproducible Example**: If the issue persists, could you please provide a minimum reproducible example? This will help us diagnose the problem more effectively. You can find guidelines for creating a reproducible example here: [Minimum Reproducible Example](https://docs.ultralytics.com/help/minimum_reproducible_example).

3. **Check Compatibility**: Verify that the versions of `ray` and `ultralytics` you are using are compatible with each other. Sometimes, specific versions of libraries have dependencies that need to be aligned.

4. **Alternative Approach**: If updating doesn't resolve the issue, you can temporarily disable the Ray Tune integration by commenting out or modifying the relevant lines in your training script.

Please try these steps and let us know if the issue persists. We're here to help!

**Comment by tiffany-sats** [comment]
hi @glenn-jocher 
for my project on object detection in a warehouse, where I used yolov8 to train my model. I want to deploy it in my MLOps pipeline, what is the general frequency for retraining a computer vision project? how do we measure accuracy of model over time? or is there a metric we should follow?

**Comment by balablb** [comment]
Hello, if I want to train at a resolution of 5K and imgsz selects 640, will Yolo change the 5K resolution to 640, or will Yolo automatically and randomly crop the original image to a 640 size for training? If I want to directly train Yolo on 5K resolution images, how should I operate? @glenn-jocher 

> Hi [@tiffany-sats](https://github.com/tiffany-sats),
> 
> Great questions! Deploying a YOLOv8 model in an MLOps pipeline for object detection in a warehouse is an exciting project. Here are some insights:
> 
> ### Retraining Frequency
> The frequency of retraining your model depends on several factors:
> 
> * **Data Drift**: If the environment or objects in the warehouse change frequently, you might need to retrain more often.
> * **Performance Degradation**: Monitor the model's performance metrics. If you notice a decline, it might be time to retrain.
> * **New Data**: Incorporate new labeled data periodically to keep the model up-to-date.
> 
> A common practice is to retrain the model every few months or whenever significant changes in the data are observed.
> 
> ### Measuring Accuracy Over Time
> To measure the accuracy of your model over time, you can use the following metrics:
> 
> * **Mean Average Precision (mAP)**: This is a standard metric for object detection models.
> * **Precision and Recall**: These metrics help you understand the balance between false positives and false negatives.
> * **F1 Score**: This combines precision and recall into a single metric.
> 
> ### Monitoring and Evaluation
> * **Validation Set**: Keep a separate validation set that is representative of your deployment environment. Evaluate the model on this set regularly.
> * **Automated Monitoring**: Implement automated monitoring in your MLOps pipeline to track these metrics continuously.
> 
> Here's a simple example of how you might set up periodic evaluation in Python:
> 
> from ultralytics import YOLO
> 
> # Load your trained model
> model = YOLO("path/to/your/model.pt")
> 
> # Define a function to evaluate the model
> def evaluate_model():
>     results = model.val(data="path/to/validation/data.yaml")
>     print(f"mAP: {results.box.map:.4f}, Precision: {results.box.precision:.4f}, Recall: {results.box.recall:.4f}")
> 
> # Schedule this function to run periodically
> import schedule
> import time
> 
> schedule.every().month.do(evaluate_model)
> 
> while True:
>     schedule.run_pending()
>     time.sleep(1)
> Feel free to adjust the schedule and metrics according to your specific needs. Best of luck with your deployment! 🚀
> 
> If you have any further questions or need additional details, don't hesitate to ask.

Hello, if I want to train at a resolution of 5K and imgsz selects 640, will Yolo change the 5K resolution to 640, or will Yolo automatically and randomly crop the original image to a 640 size for training? If I want to directly train Yolo on 5K resolution images, how should I operate? @glenn-jocher

**Comment by glenn-jocher** [comment]
@balablb when you train with 5K resolution images but set `imgsz=640`, YOLO will resize (not crop) your 5K images down to 640x640 while maintaining aspect ratio, adding padding as needed. This is the standard preprocessing behavior.

To train directly on 5K resolution images, you would set `imgsz` to a higher value like `imgsz=5120`, but this requires enormous computational resources and GPU memory that may not be feasible. A more practical approach would be to use an intermediate resolution like `imgsz=1280` or `imgsz=1920` depending on your hardware capabilities.

You can also consider [image tiling for small objects](https://docs.ultralytics.com/guides/model-evaluation-insights/#image-tiling-for-small-objects) where you divide your 5K images into smaller segments during training, which can help maintain detail while being computationally manageable.

**Comment by balablb** [comment]
> [@balablb](https://github.com/balablb) when you train with 5K resolution images but set `imgsz=640`, YOLO will resize (not crop) your 5K images down to 640x640 while maintaining aspect ratio, adding padding as needed. This is the standard preprocessing behavior.
> 
> To train directly on 5K resolution images, you would set `imgsz` to a higher value like `imgsz=5120`, but this requires enormous computational resources and GPU memory that may not be feasible. A more practical approach would be to use an intermediate resolution like `imgsz=1280` or `imgsz=1920` depending on your hardware capabilities.
> 
> You can also consider [image tiling for small objects](https://docs.ultralytics.com/guides/model-evaluation-insights/#image-tiling-for-small-objects) where you divide your 5K images into smaller segments during training, which can help maintain detail while being computationally manageable.

Thanks for the details!
I see that during inference we can use SAHI to handle very large images. Is there a similar effective solution for training, or do we have to manually crop the images offline before training?

**Comment by glenn-jocher** [comment]
@balablb for training with very large images, there isn't a built-in automated solution like SAHI integrated into the training pipeline. You would typically need to manually preprocess your 5K images by cropping them into smaller segments offline before training, and importantly, you'll need to adjust the corresponding label files to match the new cropped coordinates.

This manual preprocessing approach allows you to maintain high resolution details while keeping the training computationally feasible, but it does require additional preparation work compared to SAHI's seamless inference-time processing.
